### DISCUSSION TITLE
Any way to convert AnySchedulerOf<RunLoop> back to RunLoop?

### DISCUSSION THREAD

**Discussion by nsoojin** [pointfreeco/combine-schedulers#41] [Q&A] [answered]
**Describe the bug**
Not a bug
Hi guys, I'm using CADisplayLink in my project and want to inject AnyScheduler to test the behavior. However, CADisplayLink takes in RunLoop instead of Scheduler as a parameter to add(to:,forMode:) method. 
So I cannot use AnySchedulerOf<RunLoop> that I injected. 

**Is there a way to convert `RunLoop.main.eraseToAnyScheduler()` back to `RunLoop` type? Is it possible?**

**To Reproduce**
-

```swift
// And/or enter code that reproduces the behavior here.
private let mainLoop: AnySchedulerOf<RunLoop>

let displayLink = CADisplayLink(target: self, selector: #selector(update))
displayLink.add(to: self.mainLoop, forMode: .default) // <- Error Above
```

**Expected behavior**
-

**Screenshots**
<img width="1073" alt="Screen Shot 2021-09-07 at 12 32 11 PM" src="https://user-images.githubusercontent.com/3890050/132280186-dc68b6fc-044b-4e56-afd0-ebc1ecf0490c.png">

**Environment**
 - Xcode 12.5.1
 - Swift 5.3
 - OS (if applicable): [e.g. iOS 13]

**Additional context**
Add any more context about the problem here.

**Comment by mbrandonw** [comment]
Hi @nsoojin, unfortunately this is not possible and for good reason.

If you have an instance of `AnySchedulerOf<RunLoop>` it may not necessarily be powered by an actual run loop under the hood. In tests you may want to substitute in an immediate scheduler or a test scheduler, and so there wouldn't be a `RunLoop` to extract out of the `AnyScheduler`.

However, one of the main reasons one uses `AnyScheduler`s is for testability, but the moment you use a `CADisplayLink` you are forgoing testability. Because that API takes a concrete run loop it is just impossible to control the flow of time like you can with `TestScheduler`. 

You can see the same problem with `Timer.publisher` that ships with Foundation. It takes a concrete run loop, and so therefore is not testable at all. It's for this reason that this library ships its own [`Publishers.Timer`](https://github.com/pointfreeco/combine-schedulers/blob/main/Sources/CombineSchedulers/Timer.swift) which works with any scheduler, so that one can write deterministic tests.

If you share some details how you plan on using `CADisplayLink` in your code base we may be able to give you some hints on how you can make it a bit more testable.

You may also be interested in this similar discussion: https://github.com/pointfreeco/combine-schedulers/discussions/36

Also I'm going to convert this issue to a discussion since it's not a bug with the library.

**Reply by nsoojin** [reply]
Hey @mbrandonw, thanks for a very detailed response. It makes sense why it can't be done.

I use CADisplayLink to implement a type of 'video replay' in my app(in development). I have a sequence of images that needs to be displayed according each image's timestamp. The data structure is like this.
```swift
struct ReplayLog {
  let timestamp: TimeInterval
  let frameID: String // a unique token to retrieve the frame from a fetcher.
}
```
So I'm running a display link on main runloop, then for each cycle, I render the frame that needs to be shown based on the timestamp. 

In hopes of writing deterministic tests that don't need waiting some arbitrary amount of time, I wanted to test if I pass Array<ReplayLog> to my player, it requests the correct number of frames to the fetcher. For example, if I pass an array of three items, test if the player requested for three correct frameIDs.

**Comment by mbrandonw** [comment, marked as answer]
I think your only option for making that code testable is to design a dependency wrapper so that you can hide the `CADisplayLink`. Then the live version of the dependency can use `CADisplayLink` and the mock/test-friendly version can just use a plain timer that works with schedulers.

**Reply by nsoojin** [reply]
thanks for the guidance
